File: usaco第2章/usaco2.4/2.4c++版/maze1.cpp
/*
ID: silentf1
PROG: maze1
LANG: C++
*/
#include<fstream>
using namespace std;
ofstream fout ("maze1.out",ios::out);
ifstream fin ("maze1.in",ios::in);
int w,h,xmin[210][100],dl[20000][2],tail=0,fx[4][2]={{-2,0},{0,2},{2,0},{0,-2}},fy[4][2]={{-1,0},{0,1},{1,0},{0,-1}};
char map[210][100];
void  bfs(int head)
{
	int total=0,x,y,p,q;
	xmin[dl[1][0]][dl[1][1]]=1;
	xmin[dl[2][0]][dl[2][1]]=1;
	if((dl[tail][1]==dl[tail-1][1])&&(dl[tail-1][0]==dl[tail][0])) tail--;//相同地点只扩展一个
	while(head<=tail)
	{
		head++;
		x=dl[head][0];
		y=dl[head][1];
		for(int i=0;i<=3;i++)
		{
			p=x+fx[i][0];q=y+fx[i][1];
			if((p>0)&&(p<h)&&(q>0)&&(q<w)&&(map[x+fy[i][0]][y+fy[i][1]]==' '))  //边界条件
			if((xmin[p][q]==0)||(xmin[p][q]>(xmin[x][y]+1)))  //扩展条件
			{
				tail++;
				dl[tail][0]=p;
				dl[tail][1]=q;
				xmin[p][q]=xmin[x][y]+1;
			}
		}
	}
	for(int i=1;i<=h;i++) 
		for(int j=1;j<=w;j++)
			total=(xmin[i][j]>total?xmin[i][j]:total);
	fout<<total<<endl;
}
int main()
{
	fin>>w>>h;
	h=2*h+1;
	w=2*w+1;
	for(int i=1;i<=h;i++)
	{
		fin.get();	//读回车
		for(int j=1;j<=w;j++)
		{
			map[i][j]=fin.get();
			if((i==1)&&(map[i][j]==' ')) {tail++;dl[tail][0]=i+1;dl[tail][1]=j;}
			if((i==h)&&(map[i][j]==' ')) {tail++;dl[tail][0]=i-1;dl[tail][1]=j;}
			if((j==1)&&(map[i][j]==' ')) {tail++;dl[tail][0]=i;dl[tail][1]=j+1;}
			if((j==w)&&(map[i][j]==' ')) {tail++;dl[tail][0]=i;dl[tail][1]=j-1;}
		}
	}
	bfs(0);
    return 0;
}

File: usaco第2章/usaco2.4/2.4pascal版/maze1.pas
{
ID: sacredf1
PROG: maze1
LANG: PASCAL
}
program test;
type node=array[-1..500,-1..500]of longint;
var w,h,max,tail,head:longint;
	map:array[-1..500,-1..500]of char;
	dlx,dly:array[-1..10000]of integer;
	gt:node;
procedure main(a,b,a2,b2:longint;var g:node);//floodfill,多进程bfs
var x,y:longint;
begin
	x:=0;y:=0;
	tail:=2;head:=0;
	dlx[1]:=a;
	dly[1]:=b;//出口最先入队
	dlx[2]:=a2;
	dly[2]:=b2;//出口最先入队
	g[a div 2,b div 2]:=1;
	g[a2 div 2,b2 div 2]:=1;
	while head<=tail do
	begin
		inc(head);//头指针出队,读取新的节点.
		x:=dlx[head];
		y:=dly[head];
		//expand,扩展状态
		if (map[x+1,y]<>'-')and((x+2)<2*h+1)and(g[(x+2) div 2,y div 2]=0) then 
		begin
			inc(tail);
			dlx[tail]:=x+2;
			dly[tail]:=y;
			g[(x+2) div 2,y div 2]:=g[x div 2,y div 2]+1;
		end;
		if (map[x-1,y]<>'-')and((x-2)>1)and(g[(x-2) div 2,y div 2]=0) then
		begin
			inc(tail);
			dlx[tail]:=x-2;
			dly[tail]:=y;
		g[(x-2) div 2,y div 2]:=g[x div 2,y div 2]+1;
		end;
		if (map[x,y+1]<>'|')and((y+2)<2*w+1)and(g[x div 2,(y+2) div 2]=0) then
		begin
			inc(tail);
			dlx[tail]:=x;
			dly[tail]:=y+2;
		g[x div 2,(y+2) div 2]:=g[x div 2,y div 2]+1;
		end;
		if (map[x,y-1]<>'|')and((y-2)>1)and(g[x div 2,(y-2) div 2]=0) then
		begin
			inc(tail);
			dlx[tail]:=x;
			dly[tail]:=y-2;
		g[x div 2,(y-2) div 2]:=g[x div 2,y div 2]+1;
		end;
	end;
end;
procedure init;
var i,j:integer;
begin
	readln(w,h);
	//初始化图
	for i:=1 to h*2+1 do
	begin
		for j:=1 to w*2+1 do
			read(map[i,j]);
		readln;
	end;
end;
procedure m1;
var i,j,s,a,b,a2,b2:integer;
begin
	//flood fill
	i:=0;s:=0;
	repeat
		if i=0 then i:=1 else i:=2*h+1;
		for j:=1 to w*2+1 do
		if map[i,j]=' ' then 
		begin
			inc(s);
			case s of
			1:if (i=1) then begin a:=i+1;b:=j;end
			else begin a:=i-1;b:=j;end;
			2:if (i=1) then begin a2:=i+1;b2:=j;end
				else begin a2:=i-1;b2:=j;end;
			end;
		end;
	until i=2*h+1;
	j:=0;
	repeat
		if j=0 then j:=1 else j:=2*w+1;
		for i:=1 to h*2+1 do
		if map[i,j]=' ' then 
		begin
			inc(s);
			case s of
			1:if j=1 then begin a:=i;b:=j+1;end
				else begin a:=i;b:=j-1;end;
			2:if j=1 then begin a2:=i;b2:=j+1;end
				else begin a2:=i;b2:=j-1;end;
			end;
		end;
	until j=2*w+1;
	fillchar(gt,sizeof(gt),0);
	main(a,b,a2,b2,gt);
	max:=0;
	for i:=1 to h do
		for j:=1 to w do
			if max<gt[i,j] then max:=gt[i,j];
	writeln(max);
end;
begin
  assign(input,'maze1.in');
  reset(input);
  assign(output,'maze1.out');
  rewrite(output);
  init;
  m1;
  close(input);
  close(output);
end.